proxy.h

```diff
@@ -92,39 +91,92 @@ struct contains_traits<T, T, Us...> : applicable_traits {};
 template <class T, class U, class... Us>
 struct contains_traits<T, U, Us...> : contains_traits<T, Us...> {};
 
+template <class T, class U> struct flattening_traits_impl;
+template <class T>
+struct flattening_traits_impl<std::tuple<>, T> { using type = T; };
+template <class T, class... Ts, class U>
+struct flattening_traits_impl<std::tuple<T, Ts...>, U>
+    : flattening_traits_impl<std::tuple<Ts...>, U> {};
+template <class T, class... Ts, class... Us>
+    requires(!contains_traits<T, Us...>::applicable)
+struct flattening_traits_impl<std::tuple<T, Ts...>, std::tuple<Us...>>
+    : flattening_traits_impl<std::tuple<Ts...>, std::tuple<Us..., T>> {};
+template <class T> struct flattening_traits { using type = std::tuple<T>; };
+template <>
+struct flattening_traits<std::tuple<>> { using type = std::tuple<>; };
+template <class T, class... Ts>
+struct flattening_traits<std::tuple<T, Ts...>> : flattening_traits_impl<
+    typename flattening_traits<T>::type,
+    typename flattening_traits<std::tuple<Ts...>>::type> {};
+
 template <class... U> struct default_traits { using type = void; };
 template <class T> struct default_traits<T> { using type = T; };
 
-template <class D, class Args>
-struct dispatch_traits_impl : inapplicable_traits {};
-template <class D, class... Args>
-struct dispatch_traits_impl<D, std::tuple<Args...>> : applicable_traits {
-  using dispatcher_type = typename D::return_type (*)(const char*, Args...);
+template <class O> struct overload_traits : inapplicable_traits {};
+template <class R, class... Args>
+struct overload_traits<R(Args...)> : applicable_traits {
+  using dispatcher_type = R (*)(const char*, Args...);
+  template <class T> struct resolver { T operator()(Args...); };
 
-  template <class T>
+  template <class D, class T>
   static constexpr bool applicable_operand = requires(T operand, Args... args)
       { { D{}(std::forward<T>(operand), std::forward<Args>(args)...) }; };
-  template <class P>
-  static typename D::return_type dispatcher(const char* p, Args... args) {
+  template <class D, class P>
+  static R dispatcher(const char* p, Args... args) {
     return D{}(**reinterpret_cast<const P*>(p), std::forward<Args>(args)...);
   }
 };
+
+template <class Os, class Is> struct dispatch_traits_overload_resolution_impl;
```

Comment:
nit: `inapplicable_traits`?

Reply:
Not inheriting `inapplicable_traits` is by design. The responsibility of asserting applicability in `dispatch_traits` is on `dispatch_traits_impl` rather than `dispatch_traits_overload_resolution_impl`.